### PR TITLE
perf(expr): add initial microbenchmark for expressions

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR introduces an initial microbenchmark for expressions.

Here are the bench results of some simple operations on two I32 chunks with size 1024:

```
expr/add/i32            time:   [3.3661 µs 3.3726 µs 3.3803 µs]
expr/mul/i32            time:   [3.4544 µs 3.4628 µs 3.4707 µs]
expr/eq/i32             time:   [3.7170 µs 3.7255 µs 3.7346 µs]
expr/sum/i32            time:   [1.2756 µs 1.2800 µs 1.2860 µs]
```

... and these are the corresponding results on raw `Vec<i32>`s:

```
raw/add/i32             time:   [89.131 ns 89.174 ns 89.231 ns]
raw/sum/i32             time:   [54.860 ns 54.937 ns 55.032 ns]
```

The comparison looks really terrible. Even considering that our evaluation has null and overflow check, while the latter does not, the gap shouldn't be as large as this. This result shows that our implementation is far from optimal and has a lot of room for improvement.

To further show where the overhead comes from, this PR introduces more benches based on the raw add:

```
raw/add/i32             time:   [91.600 ns 93.113 ns 94.833 ns]
raw/add/i32/zip_eq      time:   [585.07 ns 586.54 ns 588.01 ns]
raw/add/Option<i32>/zip_eq
                        time:   [938.88 ns 940.36 ns 942.10 ns]
raw/add/Option<i32>/zip_eq,checked
                        time:   [2.1017 µs 2.1104 µs 2.1201 µs]
raw/add/Option<i32>/zip_eq,cast,checked
                        time:   [2.3200 µs 2.3598 µs 2.4268 µs]
raw/add/Option<i32>/zip_eq,checked,cast,collect_array
                        time:   [3.1290 µs 3.1444 µs 3.1605 µs]
expr/add/i32            time:   [3.3661 µs 3.3726 µs 3.3803 µs]
```

We can see that the main costs come from:
- zip_eq
- option
- error handling
- building array

~~What blows my mind is the cost from the identical type cast `i32.try_into::<i32>()`. It seems that the compiler did not realize it is infallible and then eliminate the error handling process.~~ On the other hand, when there was pure computation, the compiler could auto-vectorize it with SIMD. That's why the raw add on `Vec<i32>` performed super fast.

This analysis points out the direction of next step optimization: for primitive types, we can first apply the operation on all values without caring about null values or overflow. The benefits from SIMD are much more than saving computations from null values or errors. This conclusion has also been proven in RisingLight: https://github.com/risinglightdb/risinglight/pull/700.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#3524

### CODE REVIEW

**Comment by BugenZhao** [comment]
> for primitive types, we can first apply the operation on all values without caring about null values or overflow. The benefits from SIMD are much more than saving computations from null values or errors

I was once thinking this can be done by the compiler. 🥵 However, considering the control flow (side effect) introduced by `checked`, it sounds reasonable for not being optimized.

File: src/expr/benches/expr.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+            itertools::Itertools::zip_eq(a.iter(), b.iter())
+                .map(|(a, b)| match (a, b) {
+                    (Some(a), Some(b)) => Some(a + b),
+                    _ => None,
+                })
+                .collect::<Vec<_>>()
+        })
+    });
+    // ~2100ns
+    c.bench_function("raw/add/Option<i32>/zip_eq,checked", |bencher| {
```
Comment (BugenZhao) [inline src/expr/benches/expr.rs:118]:
```suggestion
    c.bench_function("raw/add/Option<i32>/zip,checked", |bencher| {
```

File: src/expr/benches/expr.rs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+                .zip(b.iter())
+                .map(|(a, b)| match (a, b) {
+                    (Some(a), Some(b)) => a.checked_add(*b).ok_or(Overflow).map(Some),
+                    _ => Ok(None),
+                })
+                .try_collect::<_, Vec<_>, Overflow>()
+        })
+    });
+    // ~2400ns
+    c.bench_function("raw/add/Option<i32>/zip_eq,checked,cast", |bencher| {
```
Comment (BugenZhao) [inline src/expr/benches/expr.rs:134]:
```suggestion
    c.bench_function("raw/add/Option<i32>/zip,checked,cast", |bencher| {
```

File: src/expr/benches/expr.rs (hunk 1 above)
Comment (wangrunji0408) [reply src/expr/benches/expr.rs:118]: This is intended. `zip_eq` introduced a check for each yield which greatly hurts the performance. We should switch back to `zip` and check the length at the beginning. 🤪

Comment (wangrunji0408) [reply src/expr/benches/expr.rs:118]: Sorry my fault. It should use. 🥵

File: src/expr/benches/expr.rs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        let a = (0..CHUNK_SIZE as i32).map(Some).collect::<Vec<_>>();
+        let b = (0..CHUNK_SIZE as i32).map(Some).collect::<Vec<_>>();
```
Comment (liurenjie1024) [inline src/expr/benches/expr.rs:135]: We should put these into sth like preparation?

Comment (wangrunji0408) [reply src/expr/benches/expr.rs:135]: They are now. Only the code inside `bencher.iter(...)` is benchmarked.

**Review by lmatz — APPROVED** [review]
LGTM!

**Review by BugenZhao — APPROVED** [review]

**Pull request merged**